relay-server/src/utils/squeue.rs

```diff
@@ -0,0 +1,180 @@
+use std::collections::BinaryHeap;
+use std::fmt;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll, Waker};
+use std::time::Duration;
+
+use futures::stream::FusedStream;
+use tokio::time::Instant;
+
+use futures::Stream;
+
+/// A scheduled queue that can be polled for when the next item is ready.
+#[derive(Debug)]
+pub struct ScheduledQueue<T> {
+    inner: BinaryHeap<Item<T>>,
+    waker: Option<Waker>,
+    sleep: Pin<Box<tokio::time::Sleep>>,
+}
+
+impl<T> ScheduledQueue<T> {
+    /// Creates a new, empty [`ScheduledQueue`].
+    pub fn new() -> Self {
+        Self {
+            inner: Default::default(),
+            waker: None,
+            sleep: Box::pin(tokio::time::sleep(Duration::MAX)),
+        }
+    }
+
+    /// Returns the current size of the queue.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns true if there are no items in the queue.
+    #[cfg_attr(not(test), expect(dead_code))]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Schedules a new item to be yielded at `when`.
+    pub fn schedule(&mut self, when: Instant, value: T) {
+        self.inner.push(Item { when, value });
+        if let Some(ref waker) = self.waker {
+            waker.wake_by_ref();
+        }
+    }
+}
+
+impl<T> Unpin for ScheduledQueue<T> {}
+
+impl<T> Stream for ScheduledQueue<T> {
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.waker = Some(cx.waker().clone());
+
+        let sleep = self.sleep.as_mut().poll(cx);
+
+        // No item means, we're pending until scheduled (through adding an item).
+        if let Some(next) = self.inner.peek() {
+            let when = next.when;
+
+            // TODO: maybe optimize here for `when` being in the past,
+            // currently self goes through one additional wakeup and sleep poll.
```

Comment:
Can you expand on this? I believe if you hit this case, `when` will (practially) always be in the past because `when <= Instant::now()`, right?

Reply:
Rewrote the comment to include a lot more context, hopefully this is now better explained.

Reply:
That's pretty much what the todo is about, I wrote this over a year ago for a personal project so details are a bit blurred. But from what I remember is I compromised on this to have the least amount of possible branches.

1. Always poll the sleep to make sure the waker is registered
2. If the sleep is ready, check if we can yield the item
3. If the sleep does not match the current item, update/reset the sleep deadline and wait for another wake up

Now the version here is slightly different because it has no way of modifying the heap while it is being polled, my private version has interior mutability on the heap, allowing concurrent modifications, which leads to a few more edge cases that are still handled correctly with this control flow.